### PR TITLE
Array/Object transform canonization

### PR DESCRIPTION
**Pull request by Darrellbor**
PR for issue #297 
Implement canonization for 
- [x] Array Transform Canonization
- [x] Object Transform Canonization

### CODE REVIEW

**Comment by Darrellbor** [comment]
@gavv completed and ready for review

**Comment by gavv** [comment]
Thanks for PR!

There is one minor issue. In this test, we want to check how Transform performs canonization, and we do NOT want to check how IsEqual performs canonization.

So instead of calling IsEqual with various types, we should call Transform with various types and then check that newArray.Raw() always returns canonized value. We don't need IsEqual at all in this test.

**Comment by Darrellbor** [comment]
@gavv I have adjusted the array transform to `assert` equals rather than direct equals; although i can no longer use wrapped types like 
```
type (
 myArray []interface{}
 myFloat float64
)
```
As it complains that `float64` does not equal `myFloat` for example.

Do i also do the same for object transform?

**Comment by gavv** [comment]
Wrapped types should be returned from Transform, not passed to assert.Equal. myArray is not relevant, but myFloat can be used.

> Do i also do the same for object transform?

Yep.

File: array_test.go [hunk 1]
```diff
@@ -1582,6 +1582,35 @@ func TestArray_Transform(t *testing.T) {
...
+		newArray.chain.assertNotFailed(t)
+		newArray.chain.clearFailed()
+
+		assert.NotEqual(t, []interface{}{"2.0", float64(4), float64(6)}, newArray.Raw())
+		newArray.chain.assertNotFailed(t)
+		newArray.chain.clearFailed()
+
+		assert.Equal(t, []interface{}{2.0, float64(4), 6.0}, newArray.Raw())
+		newArray.chain.assertNotFailed(t)
+		newArray.chain.clearFailed()
```
Comment (gavv) [inline array_test.go:1612]: These three checks can be reduced to only one:

```go
		assert.Equal(t, []interface{}{2.0, 4.0, 6.0}, newArray.Raw())
		newArray.chain.assertNotFailed(t)
```

Comment (Darrellbor) [reply array_test.go]: done

**Comment by gavv** [comment]
Thanks!

**Pull request merged**